qiskit/providers/ibmq/api_v2/rest/schemas/root.py

```diff
@@ -0,0 +1,167 @@
+# -*- coding: utf-8 -*-
+
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2017, 2019.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Schemas for root."""
+
+from marshmallow.validate import OneOf
+from qiskit.providers import JobStatus
+from qiskit.providers.models.backendconfiguration import BackendConfigurationSchema
+from qiskit.providers.ibmq.apiconstants import ApiJobKind
+from qiskit.validation import BaseSchema
+from qiskit.validation.fields import String, Dict, Nested, Boolean, List, Number
+
+
+# Helper schemas.
+
+class ProjectResponseSchema(BaseSchema):
+    """Nested schema for ProjectsResponseSchema"""
+
+    # Required properties.
+    isDefault = Boolean(required=True)
+
+
+class ProjectsResponseSchema(BaseSchema):
+    """Nested schema for GroupResponseSchema"""
+
+    # Required properties.
+    project_name = String(required=True)
+    project = Nested(ProjectResponseSchema, required=True)
+
+
+class GroupResponseSchema(BaseSchema):
+    """Nested schema for GroupsResponseSchema"""
+
+    # Required properties.
+    projects = Dict(Nested(ProjectsResponseSchema), required=True)
+
+
+class GroupsResponseSchema(BaseSchema):
+    """Nested schema for HubsResponseSchema"""
+
+    # Required properties.
+    group_name = String(required=True)
+    group = Nested(GroupResponseSchema, required=True)
+
+
+class CircuitErrorResponseSchema(BaseSchema):
+    """Nested schema for CircuitResponseSchema"""
+
+    # Required properties
+    code = String(required=True, validate=OneOf(['GENERIC_ERROR', 'HUB_NOT_FOUND']))
+
+
+class BackendRequestSchema(BaseSchema):
+    """Nested schema for JobsRequestSchema"""
+
+    # Required properties
+    name = String(required=True, description="the name of the backend.")
+
+
+class JobsStatusFilterRequestSchema(BaseSchema):
+    """Nested schema for JobsStatusRequestSchema"""
+
+    # Optional properties
+    where = Dict(attribute="extra_filter", required=False,
+                 description="additional filtering passed to the query.")
+
+    # Required properties
+    order = String(required=True, default="creationDate DESC")
+    limit = Number(required=True, description="maximum number of items to return.")
+    skip = Number(required=True, description="offset for the items to return.")
+
+
+# Endpoint schemas.
+
+class HubsResponseSchema(BaseSchema):
```

Comment:
Can you double-check this schema? It seems to be missing one nesting level - the response contains a _list_ of hubs (each one containing a dict of groups, each one containing a dict of projects).

In this case, you can probably just use `many` (it being a list):

```python
hubs = Nested(SomethingAboutHubs, required=True, many=True)
```

Reply:
Exactly, that what I pointed out in the PR message:

* Deserialization of `HubsResponseSchema` will always fail bc `@bind_schema` does not allow passing `many = True`

The issue was raised under `qiskit-terra` repo: https://github.com/Qiskit/qiskit-terra/issues/3021
If we were to create a model for `HubsResponseSchema` using `@bind_schema` decorator, and then use it to deserialize hubs response payload, the deserialization will always fail with `ModelValidationError`. The reason is bc the payload is of the following format: `[<payload>]` i.e. the payload is wrapped in an array. This could be easily fixed by passing `many = True` while creating an instance of `HubsResponseSchema(many=True)`, **but** the instance is created internally in `@bind_schema` decorator, and we *cannot*  pass any arguments inside.

Reply:
Otherwise, I'd have to have *another* level of nestness. Like:
```
class HubResponseSchema(BaseSchema):
    """Schema for HubResponse"""

    # Required properties.
    name = String(required=True)
    groups = Dict(Nested(GroupsResponseSchema), required=True)

class HubsResponseSchema(BaseSchema):
    """Schema for HubsResponse"""

    # Required properties.
    huns = Nested(HubResponseSchema, required=True, many=True)
```

Reply:
So, it is either another schema *or* upgrading `@bind_schema` to support `many=True`

Reply:
@diego-plan9 Here is what I found.

Two schemas approach won't work! It will only work for payloads of the type:
```
{
    "hubs": [ 
        {<hub_1>},
        ...
    ]
}
```

But in our case, the payload is:
```
[
{<hub_1>},
...
]
```

That's why the only option is still having `@bind_schema(HubsResponseSchema, many=True)`
In order to deserialize "raw" collection.

Let me know if you need any clarification.

I'll try to create a quick code sample to reproduce this

Reply:
Here is the code sample:
```
from unittest import TestCase

from marshmallow import Schema
from marshmallow.fields import Nested, String, List



class NestedTestSchema(Schema):
    test = String(required=True)

class ManyTestSchema(Schema):
    nested = Nested(NestedTestSchema, many=True)

class ListTestSchema(Schema):
    nested = List(Nested(NestedTestSchema))

class SchemaTest(TestCase):

    def test_many_and_list_schemas_same(self):
        payload = { 'nested': [
                        {'test': 'something'}
                    ]
                  }
        many_deserialized = ManyTestSchema().load(payload).data
        list_deserialized = ListTestSchema().load(payload).data
        self.assertEqual(many_deserialized, list_deserialized)

    def test_cannot_deserialize_raw_list(self):
        payload = [
                    {'test': 'something'}
                  ]
        self.assertTrue(ManyTestSchema().load(payload).errors)
        self.assertTrue(ListTestSchema().load(payload).errors)
        self.assertTrue(ManyTestSchema(many=True).load(payload).data)
        self.assertTrue(ListTestSchema(many=True).load(payload).data)
```

For `ManyTestSchema` and `ListTestSchema` I wanted to point out that there is no difference in using either `Nested(NestedTestSchema, many=True)` or `List(Nested(NestedTestSchema))`.